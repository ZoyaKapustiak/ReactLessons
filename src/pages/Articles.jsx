import React from "react";
import { useEffect } from "react";
import { Button } from "@mui/material";
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { CardActionArea } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import Box from '@mui/material/Box';
import { getAllArticles } from '../store/articles/actions'
import { useDispatch, useSelector } from "react-redux";
import { selectArticles, selectArticlesError, selectArticlesLoading } from "../store/articles/selectors";

export function Articles() {
  
  const dispatch = useDispatch()
  const articles = useSelector(selectArticles)
  const error = useSelector(selectArticlesError)
  const loading = useSelector(selectArticlesLoading)

  const requestArticles = () => {
    dispatch(getAllArticles())
  }

  useEffect(() => {
    requestArticles()
  }, [])

  if(error) {
    return (
      <>
      <h3>Error</h3>
      <button onClick={requestArticles} >Retry</button>
      </>
    )
  }

  return (
    <>
      <h1>Article</h1>
      <Button 
        variant="contained" 
        color="success"
        onClick={requestArticles}
        >
          Get API
      </Button>
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center' }}>
          <CircularProgress />
        </Box>
      )}
      {!loading && articles.map((article) => (
      <Box sx={{ display: 'flex', justifyContent: 'center' }} key={article.id}>
        <Card sx={{ maxWidth: 345 }} key={article.id} >
          <CardActionArea>
            <CardMedia
              component="img"
              height="140"
              image={article.imageUrl}
              alt="green iguana"
            />
            <CardContent>
              <Typography gutterBottom variant="h5" component="div">
                {article.title}
              </Typography>
              <Typography variant="body2" color="text.secondary">
                {article.summary}
              </Typography>
            </CardContent>
          </CardActionArea>
      </Card>
    </Box>
    ))} 
    </>
  )
}