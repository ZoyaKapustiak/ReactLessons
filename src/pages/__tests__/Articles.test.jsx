import React from "react";
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'
import userEvent from "@testing-library/user-event";
import configureStore from 'redux-mock-store';
import { Articles } from "../Articles";
import * as reduxHooks from 'react-redux'

jest.mock('react-redux')

describe('testing Articles Page', () => {
  const mockUseSelector = jest.spyOn(reduxHooks, 'useSelector')
  const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch')
  const dispatch = jest.fn()

  it('creating snapshot Articles Pages', () => {
    mockUseSelector.mockReturnValue([])
    mockUseDispatch.mockReturnValue(dispatch)
    const component = {}
      render(
            <Articles/>     
          )
    expect(component).toMatchSnapshot()
  })
  it('testing button functioan', async () => {
    render(
      <Articles/>
    )
    const button = screen.getByRole('button')
    await userEvent.click(button)
    expect(dispatch).toHaveBeenCalled()
  })
})