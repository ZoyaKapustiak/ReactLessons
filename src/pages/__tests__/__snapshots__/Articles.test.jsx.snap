// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Articles Page creating snapshot Articles Pages 1`] = `Object {}`;
